simulation/robot_calibration.py

```diff
@@ -141,101 +416,157 @@ def _decode(self, line):
-                reset_arrays()
-                stop_timer = 0
-            elif line.startswith(self.END_PATTERN):
-                stop_timer = 1
-                break
-            else:
-                self._decode(line)
+            print('\x1b[0;35;40m' + line + '\033[0m', file=output, end='')
+            output.flush()
+            self._decode(line)
         print('Parsing finished')
 
 
 class NativeParser(Parser):
 
     def parse(self, p):
+        """
+        Read subprocess 'p' outcoming character from stdout and parse them in
+        upper class.
+        """
         super(NativeParser, self).parse(p.stdout)
 
     def send(self, p):
+        """
+        Read this python script incoming character from stdin and when a line is
+        completed, propagates the whole string to subprocess 'p' stdin stream.
+        """
         for line in fileinput.input():
-            p.stdin.write(line)
+            p.stdin.write(line.encode('utf-8'))
+            p.stdin.flush()
 
 class SerialParser(Parser):
     def __init__(self, serial_port):
+        super().__init__()
         self.ser = Serial(port=serial_port, baudrate=115200, timeout=1, writeTimeout=1)
 
     def parse(self):
-            if self.ser.isOpen():
-                while True:
-                    super(SerialParser, self).parse(self.ser)
+        """
+        Read serial port incoming character (i.e. MCU firwmare stdout) and parse
+        them it in upper class
+        """
+        if self.ser.isOpen():
+            while True:
+                super(SerialParser, self).parse(self.ser)
 
-    def send(self, p):
+    def send(self):
+        """
+        Read this python script incoming character from stdin and when a line is
+        completed, propagates the whole string to serial port.
+        """
         for line in fileinput.input():
-                if self.ser.isOpen():
-                    self.ser.write(line)
+            if self.ser.isOpen():
+                self.ser.write(bytes(line, 'utf8'))
+
+def usage(pname = ""):
+    print(f"Usage : {pname} [-h] [-d /dev/ttyUSB0]")
+    print("")
+    print("Launch PID measurement tool. Works on native board or real target through serial link.")
+    print("")
+    print("\t-h\t\tThis help message.")
+    print("\t-d, --device=\tSpecify UART device, run on native board otherwise.");
 
 if __name__ == "__main__":
+
+    # Parse command line arguments
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], "hD:", ["help", "device="])
+    except getopt.GetoptError as err:
+        print(str(err))  # will print something like "option -a not recognized"
+        usage(pname = sys.argv[0])
+        sys.exit(2)
+
+    uart_dev = None
+    for o, a in opts:
+        if o in ("-h", "--help"):
+            usage(pname = sys.argv[0])
+            sys.exit()
+        elif o in ("-D", "--device"):
+            uart_dev = a
+        else:
+            assert False, "unhandled option"
+
```

Comment:
getopt module is C-developers friendly but almost deprecated. Please consider using [argparse](https://docs.python.org/3.7/library/argparse.html#module-argparse) module instead.